### PR TITLE
RankLLM Refactor

### PR DESCRIPTION
**Pull request by xpbowler**
## Summary of Changes

**Abstract Base Classes**
- New ABC: `RankLLM` with 3 children: `ListwiseRankLLM`, `PairwiseRankLLM`, `PointwiseRankLLM`
- `RankLLM` class is simple: 3 member variable, essential reranking abstract functions
- `Reranker` class

**`ListwiseRankLLM`**
- Provided base implementation for 

**`__init__.py` Files**
- Add functions and classes to directory namespace --> easier, simpler syntax when importing
- Added for `/rerank` and `/retrieve` folders (including subfolders)

**Usage of `**kwargs**`**
- `**kwargs` is used in many places throughout the refactor to specify variable keyword argument inputs
-  `retrieve_and_rerank()` now uses **kwargs to pass in many of the variables that are used *only* under specific scenarios. This decision was made because it declutters the parameters.
- **Pros**: declutters parameters, makes parameters scalable - can add as many as would like
- **Cons**: Makes parameters ambiguous. Should probably add comments and documentation regarding exactly how to input which parameters and under which circumstances
- used in `rerank_batch()` for parameters specific to one concrete class. *For example*, the `vllm_batched` class of the `RankListwiseOSLLM` class. This parameter is only used for this one class, so it does not make sense to add the param to the abstract function inside the ABC `RankLLM`

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -1,10 +1,17 @@
...
-openai>=1.23.6
-tiktoken>=0.6.0
-transformers>=4.40.1
-pyserini>=0.35.0
-python-dotenv>=1.0.1
-faiss-cpu>=1.8.0
-ftfy>=6.2.0
-dacite>=1.8.1
-fschat[model_worker]>=0.2.36
+dacite==1.8.1
```
Comment (ronakice) [inline requirements.txt:1]: Let's go with >=, I like some flexibility and models grow too fast

File: src/rank_llm/demo/rerank_rank_gpt.py [hunk 2]
```diff
@@ -6,7 +6,7 @@
 parent = os.path.dirname(parent)
 sys.path.append(parent)
 
-from rank_llm.rerank.api_keys import get_openai_api_key
+from rank_llm.rerank.listwise.gpt.api_keys import get_openai_api_key
 from rank_llm.rerank.rank_gpt import SafeOpenai
```
Comment (ronakice) [inline src/rank_llm/demo/rerank_rank_gpt.py:10]: this path changes?

File: src/rank_llm/rerank/listwise/gpt/rank_gpt.py [hunk 3]
```diff
@@ -149,20 +180,35 @@ def run_llm(
     def _get_prefix_for_rank_gpt_prompt(
         self, query: str, num: int
     ) -> List[Dict[str, str]]:
+        # APEER
         return [
             {
                 "role": "system",
-                "content": "You are RankGPT, an intelligent assistant that can rank passages based on their relevancy to the query.",
+                "content": "You are RankGPT, an advanced assistant specialized in ranking passages by their relevance to a given query.",
```
Comment (ronakice) [inline src/rank_llm/rerank/listwise/gpt/rank_gpt.py:187]: Why is this changing? It should be what is in the main branch

File: src/rank_llm/rerank/listwise/gpt/rank_gpt.py [hunk 4]
```diff
@@ -149,20 +180,35 @@ def run_llm(
...
+        #         "role": "user",
+        #         "content": f"I will provide you with {num} passages, each indicated by number identifier []. \nRank the passages based on their relevance to query: {query}.",
+        #     },
+        #     {"role": "assistant", "content": "Okay, please provide the passages."},
+        # ]
+
     def _get_suffix_for_rank_gpt_prompt(self, query: str, num: int) -> str:
-        return f"Search Query: {query}. \nRank the {num} passages above based on their relevance to the search query. The passages should be listed in descending order using identifiers. The most relevant passages should be listed first. The output format should be [] > [], e.g., [1] > [2]. Only response the ranking results, do not say any word or explain."
+        # APEER
+        return f"Search Query: {query}.\nArrange the {num} passages above in order of relevance to the search query, from most to least relevant. Use the numerical identifiers for ranking. The format should be [] > [], e.g., [1] > [2]. Only provide the ranking results without any additional text or explanation."
```
Comment (ronakice) [inline src/rank_llm/rerank/listwise/gpt/rank_gpt.py:210]: I feel a lot of this is different in main branch, please take a look

**Review by ronakice — CHANGES_REQUESTED** [review]
I've left my first set of comments will look more after!

File: src/rank_llm/retrieve_and_rerank.py [hunk 5]
```diff
@@ -22,81 +24,112 @@ def retrieve_and_rerank(
...
+            print(f"Pass {pass_ct + 1} of {num_passes}:")
+
+            rerank_results = reranker.rerank_batch(
+                requests,
+                rank_end=top_k_retrieve,
+                rank_start=0,
+                shuffle_candidates=shuffle_candidates,
+                logging=print_prompts_responses,
+                top_k_retrieve=top_k_retrieve,
+                **kwargs,
```
Comment (sahel-sh) [inline src/rank_llm/retrieve_and_rerank.py:81]: Have you tested this cl with unexpected keys in kwargs? I expect passing kwargs here, without changing the `rerank_batch`  would simply flatten the kwargs elements as keyword parameters and raise type error.
Both rerank and rerank_batch functions should take **kwargs as input and properly filter it (similar to what you are doing in the get_reranker and retrive function here).

File: src/rank_llm/retrieve_and_rerank.py [hunk 6]
```diff
@@ -111,68 +144,143 @@ def retrieve_and_rerank(
...
+        kwargs.get(key_and_default[0], key_and_default[-1])
+        for key_and_default in keys_and_defaults
+    ]
+
+    # Check that type of provided kwarg is compatible with the provided default type
+    for i, extracted_kwarg in enumerate(extract_kwargs):
+        if type(keys_and_defaults[i[-1]]) != None and (
+            type(extracted_kwarg) != type(keys_and_defaults[i[-1]])
         ):
-            print("Evaluating:")
```
Comment (sahel-sh) [inline src/rank_llm/retrieve_and_rerank.py:171]: why are you removing these evals
[general tip]: Always aim to separate refactoring from logical changes unless the logical change is the goal of the refactoring (e.g. using a new library that introduces a new dependency, etc).

File: src/rank_llm/rerank/reranker.py [hunk 7]
```diff
@@ -1,187 +0,0 @@
...
-from pathlib import Path
-from typing import List
-
-from tqdm import tqdm
-
-from rank_llm.data import DataWriter, Request, Result
-from rank_llm.rerank.rankllm import RankLLM
-
-
-class Reranker:
```
Comment (sahel-sh) [inline src/rank_llm/rerank/reranker.py:11]: IIUC, you are removing the Reranker class and add rerank and rerank_batch to the RankLLM class. 
hm, I am not sure if we want this. I kind of like a Reranker class that has a ranking agent. For example, writing the reranking ressults can be a task of the reranker, but not the ranking agent.
WDYT? @ronakice

Comment (sahel-sh) [reply src/rank_llm/rerank/reranker.py:11]: We might also have Rerankers with other agents. For example, our Identity and Random rerankers do not inherit from rankllm but still need to implement the rerank and rerank_batch methods.

**Review by sahel-sh — COMMENTED** [review]
@xpbowler I had missed to open some of the diffs that were not automatically loaded. I resolved a couple of my previous comments, sorry about that!

File: src/rank_llm/retrieve/retriever.py [hunk 8]
```diff
@@ -65,12 +65,6 @@ def from_dataset_with_prebuilt_index(
         """
         if not dataset_name:
             raise ValueError("Please provide name of the dataset.")
-        if not isinstance(dataset_name, str):
```
[the remark this answers is not in the document]
Comment (sahel-sh) [reply src/rank_llm/retrieve/retriever.py:68]: @xpbowler PTAL

File: src/rank_llm/demo/experimental_results.py [hunk 9]
```diff
@@ -8,17 +8,13 @@
...
-from rank_llm.rerank.zephyr_reranker import ZephyrReranker
+from rank_llm.rerank import get_openai_api_key
+from rank_llm.rerank.listwise import SafeOpenai, VicunaReranker, ZephyrReranker
 from rank_llm.retrieve.retriever import Retriever
 from rank_llm.retrieve.topics_dict import TOPICS
 
 # create rerankers
-agent = SafeOpenai("gpt-3.5-turbo", 4096, keys=get_openai_api_key())
-g_reranker = Reranker(agent)
+g_reranker = SafeOpenai("gpt-3.5-turbo", 4096, keys=get_openai_api_key())
```
[the remark this answers is not in the document]
Comment (sahel-sh) [reply src/rank_llm/demo/experimental_results.py]: @xpbowler PTAL

File: src/rank_llm/retrieve_and_rerank.py [hunk 10]
```diff
@@ -22,81 +18,115 @@ def retrieve_and_rerank(
...
+        query,
+        qid,
+        dataset=dataset,
+        **kwargs,
+    )
+    print(f"Retrieval complete!")
+
+    # Reranking stage
+    print(f"Reranking and returning {top_k_rerank} passages with {model_path}...")
+    if agent is None:
```
[the remark this answers is not in the document]
Comment (sahel-sh) [reply src/rank_llm/retrieve_and_rerank.py]: Please address this @xpbowler

File: src/rank_llm/rerank/rankllm.py [hunk 11]
```diff
@@ -33,23 +22,8 @@ class RankLLM(ABC):
     def __init__(
         self,
         model: str,
-        context_size: int,
-        prompt_mode: PromptMode,
-        num_few_shot_examples: int,
     ) -> None:
         self._model = model
-        self._context_size = context_size
```
[the remark this answers is not in the document]
Comment (sahel-sh) [reply src/rank_llm/rerank/rankllm.py:41]: @xpbowler PTAL

File: src/rank_llm/rerank/reranker.py [hunk 12]
```diff
@@ -1,187 +1,285 @@
...
+        )
+        return results[0]
+
+    def write_rerank_results(
+        self,
+        retrieval_method_name: str,
+        results: List[Result],
+        shuffle_candidates: bool = False,
+        top_k_candidates: int = 100,
+        pass_ct: int = None,
```
[the remark this answers is not in the document]
Comment (sahel-sh) [reply src/rank_llm/rerank/reranker.py:96]: @xpbowler PTAL

**Review by sahel-sh — APPROVED** [review]

File: src/rank_llm/retrieve/retriever.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ronakice) [reply src/rank_llm/retrieve/retriever.py:68]: @xpbowler did you verify this?

**Review by ronakice — APPROVED** [review]
LGTM! but left one comment

**Review by ronakice — APPROVED** [review]
LGTM!

**Pull request merged**